File: 347-top-k-frequent-elements/347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        /* the best implementation of this would be to use a hashmap and treat
         * each number as a key, that way it is easier to return the top K keys
         * by using a priority queue */
        
        
           //count the frequency for each element
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        // create a min heap
        PriorityQueue<Map.Entry<Integer, Integer>> queue 
                      = new PriorityQueue<>(Comparator.comparing(e -> e.getValue()));

        //maintain a heap of size k.
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            queue.offer(entry);
            if (queue.size() > k) {
                queue.poll();
            }
        }

        //get all elements from the heap
        int[] resultArr=new int[queue.size()];
        int index =0;
        while (queue.size() > 0) {
            resultArr[index++]=queue.poll().getKey();
        }

        return resultArr;
    }
        
        
    
}

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        dic = {}
        
        for val in nums:
            if val in dic:
                dic[val] += 1
            else:
                dic[val] = 1
                
                
        num = []
        s = sorted(dic.items(), key=lambda x: x[1], reverse=True)

            
        
        print(s)
        for i in range(k):
            k = s[i][0]
            num.append(k)
        return num
            